src/recovery/mod.rs

```diff
@@ -1254,35 +1274,31 @@ where
 
                     // Now let's find out which epochs we should wake
                     // up the logic for.
-                    assert!(activate_epochs_buffer.is_empty());
-
-                    // Try to process all the epochs we have data for.
-                    // Filter out epochs that are ahead of the
-                    // frontier. The state at the beginning of those
-                    // epochs are not truly known yet, so we can't
-                    // apply input in those epochs yet.
-                    activate_epochs_buffer.extend(
+                    assert!(closed_epochs_buffer.is_empty());
+
+                    // On the last activation, we eagerly executed the
```

Comment:
This comment confuses me in conjunction with the comment below (on 1296). At this point in the code, have we already executed the frontier? Or are we about to, like in 1298?

Reply:
Possibly both! If the frontier hasn't changed between activations, we keep running it multiple times merging in new data in that singular frontier epoch until it is closed.

E.g: We have a single stateful operator. It has a single input.

1. Wake up. Poll the input. It returns some data in epoch 4, 5, 6. The input's frontier is epoch 6. That means epoch 6 is _not_ closed. But 4,5 < 6 so they are closed so process them in order. Also process 6! 6 isn't closed but by definition we won't see data before it ever again, so we can partially complete processing on it. Go to sleep.
2. Wake up. Poll the input. It returns more data in epoch 6. The input's frontier is still 6. Process 6 again; not because it is closed, but because we're still "in order" and we can partially complete work. Go to sleep.

So in that case we eagerly executed the same frontier on both awakenings.

---

But also the frontier might have changed between activations. In that case, the answer to the question is still "both" but we already executed the old frontier and we're about to execute the new frontier. The exact epoch of "frontier" has changed.

---

Perhaps it would be clarifying here to put "the frontier on the last activation" and in the next comment to put "the current frontier" (and those might be the same and that's ok).